### PR TITLE
Support Distributing Lock Statement

### PR DESCRIPTION
**Pull request by gledis69**
DESCRIPTION: 
* Lock statements will propagate locks to all the nodes in the metadata when a citus tables or a view is locked. Locks will also be propagated for all citus tables that appear in the definition of views that are locked (recursively applies to the views inside the locked view).
* TRUNCATE-ing a citus table from a worker node is no longer allowed if the coordinator is not added to the metadata. When TRUNCATE is called on a citus table that table needs to be locked in all the nodes before the TRUCATE is executed. If the coordinator is not in the metadata, the coordinator will not be aware of the lock on the table and can access the table's shard while a TRUNCATE is executing. Despite not being recommended, this behavior can be allowed by setting `SET citus.allow_unsafe_locks_from_workers TO 'on';`.
* `pg_dump` which calls LOCK TABLE is affected in the same way as TRUNCATE. An error is thrown when locking from worker when coordinator is not in the metadata.

### CODE REVIEW

File: src/backend/distributed/utils/resource_lock.c [hunk 1]
```diff
@@ -1196,3 +1199,147 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+		return;
+	}
+
+	List *distributedRelationList = NIL;
+
+	/*
+	 * this is used to enforce the lock order:
+	 * [...A], [...B]
+	 * Where:
+	 * A: Excplicitly locked relations
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1229]: minor: typo Excplicitly


Also, `Relations which depend on` -> maybe `Relations that has referencing foreign key relationship` or such

File: src/backend/distributed/utils/resource_lock.c [hunk 2]
```diff
@@ -1196,3 +1199,147 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+		{
+			CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
+			Assert(cacheEntry != NULL);
+
+			List *referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+
+			Oid referencingRelationId = InvalidOid;
+			foreach_oid(referencingRelationId, referencingTableList)
+			{
+				dependentRelationIds = lappend_oid(dependentRelationIds,
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1299]: what if `referencingRelationId` returns false for `ShouldSyncTableMetadata`? Maybe we can add that check as well

File: src/backend/distributed/shared_library_init.c [hunk 3]
```diff
@@ -670,6 +670,21 @@ RegisterCitusConfigVariables(void)
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.allow_unsafe_locks_from_workers",
+		gettext_noop("Enables acquiring a distributed lock from a worker "
+					 "when the coordinator is not in the metadata"),
+		gettext_noop("Set to false by default. If set to true, enables "
```
Comment (onderkalaci) [inline src/backend/distributed/shared_library_init.c:677]: maybe briefly touch why? Because it might cause concurrent operations from the coordinator and the workers do not follow the lock ordering.

File: src/backend/distributed/utils/resource_lock.c [hunk 4]
```diff
@@ -1081,14 +1083,21 @@ EnsureCanAcquireLock(Oid relationId, LOCKMODE lockMode)
...
+ * AcquireDistributedLockOnRelations_Internal acquire a distributed lock on worker nodes
  * for given list of relations ids. Worker node list is sorted so that the lock
- * is acquired in the same order regardless of which node it was run on. Notice
- * that no lock is acquired on coordinator node if the coordinator is not added
- * to the metadata.
+ * is acquired in the same order regardless of which node it was run on.
+ * Notice that no lock is acquired on coordinator node if the coordinator is not
+ * added to the metadata.
+ *
+ * Notice that no validation is done on the relationIds, that is the responsibility of
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1092]: AcquireDistributedLockOnRelations (or any caller in general)

File: src/backend/distributed/utils/resource_lock.c [hunk 5]
```diff
@@ -1104,22 +1113,14 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
 	 */
 	bool startedLockCommand = false;
 
+	/* Used to terminate partial lock commands */
+	const char *lockTerminationString = nowait ? " IN %s MODE NOWAIT;\n" :
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1117]: aren't we sure that no code-path can ever pass `nowait` to `lock_relation_if_exists`? I mean,
- We only use LOCK_RELATION_IF_EXISTS for foreign tables
- Foreign TABLEs do NOT support LOCK command:
```SQL
BEGIN;
LOCK foreign_table IN SHARE MODE;
ERROR:  "foreign_table" is not a table or view
Time: 0.292 ms
```

So, my suggestion would be to throw an error if `nowait` is true?

With that, we can avoid the changes on the LOCK_RELATION_IF_EXISTS, and avoid confusion. Also, when that is supported, we can remove the error check and implement what you implement now. But, seems not necessary at this point

File: src/test/regress/expected/multi_mx_hide_shard_names.out [hunk 6]
```diff
@@ -74,15 +74,14 @@ SELECT * FROM citus_shard_indexes_on_worker WHERE "Schema" = 'mx_hide_shard_name
 -- make sure that pg_class queries do not get blocked on table locks
 begin;
 lock table test_table in access exclusive mode;
-prepare transaction 'take-aggressive-lock';
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_mx_hide_shard_names.out:77]: Instead of changing this, we should do:

```SQL
BEGIN;
SET LOCAL citus.enable_ddl_propagation TO OFF;
lock table test_table in access exclusive mode;
prepare transaction 'take-aggressive-lock';

File: src/backend/distributed/shared_library_init.c [hunk 7]
```diff
@@ -670,6 +670,21 @@ RegisterCitusConfigVariables(void)
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.allow_unsafe_locks_from_workers",
```
Comment (marcocitus) [inline src/backend/distributed/shared_library_init.c:674]: is there any reason why we would want to allow this?

It seems better to throw an error telling users to add the coordinator to the metadata, since we don't really want to support clusters without it.

**Review by onderkalaci — COMMENTED** [review]
In general looks pretty reasonable. Definitely needs tests, I guess you'd move the tests you wrote in the earlier PR? It is always good to check PG regression tests for the same feature (and potentially adopt similar tests: https://github.com/postgres/postgres/blob/master/src/test/regress/sql/lock.sql)


I'll test a bit more as well, wanted to share my comments regarding the code

File: src/backend/distributed/shared_library_init.c (hunk 7 above)
Comment (onderkalaci) [reply src/backend/distributed/shared_library_init.c:674]: Mostly not to break existing users who only queries from workers and keep regression test changes manageable. Otherwise we end with a quite large PR -- even comparable to this one: https://github.com/citusdata/citus/pull/5756. I think once we implement 5756, we could drop this GUC as well?

Comment (marcocitus) [reply src/backend/distributed/shared_library_init.c:674]: Aaah, for truncate, got it.

**Review by onderkalaci — CHANGES_REQUESTED** [review]
It seems we ignore `only` keyword, see here: https://www.postgresql.org/docs/current/sql-lock.html

 Lets make sure that we fix merging.

also `*`: Optionally, * can be specified after the table name to explicitly indicate that descendant tables are included. From pg docs: https://www.postgresql.org/docs/current/sql-lock.html

File: src/backend/distributed/shared_library_init.c (hunk 7 above)
Comment (onderkalaci) [reply src/backend/distributed/shared_library_init.c:674]: Oh, yes, the primary goal is TRUNCATE. Well, thanks for commenting on this, now the reason is documented on Github as well:)

File: src/backend/distributed/utils/resource_lock.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/resource_lock.c:1117]: (same for ONLY and * for now)

Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1117]: There is no code path that triggers `lock_relation_if_exists(..., nowait => true)` unless the user explicitly calls that.
There are a few things that can be done about this:
1. Remove the nowait support from `lock_relation_if_exists` since we are not using it anyway.
2. Throw an error inside `lock_relation_if_exists` if `nowait` is true. I think this is counterintuitive to the udf since what is the point of having a boolean parameter if one of its values is disallowed.
3. Throw an error inside `AcquireDistributedLockOnRelations_Internal` when a foreign table is being locked with `nowait` true. Not sure what is appropriate for the error to say.

Let me know what you think is best.

File: src/backend/distributed/utils/resource_lock.c (hunk 4 above)
Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1092]: I will generalize this to specify any caller.

File: src/backend/distributed/utils/resource_lock.c (hunk 2 above)
Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1299]: My reasoning was that citus tables need to have other citus tables as dependencies, but this check does not hurt to have us covered for potential future uses of the function.

File: src/backend/distributed/utils/resource_lock.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/resource_lock.c:1117]: I think we can do (1) and (3) together. The error could be `this feature is currently not implemented` or such. As only Citus devs would be able to trigger the error when they refactor things to hit the error, I wouldn't worry about much

**Comment by gledis69** [comment]
https://github.com/citusdata/citus/pull/5930#discussion_r870045304
This change of behavior for TRUNCATE can be suppressed with this change:
https://github.com/citusdata/citus/blob/b2db5c1f44d146caca68610b495e4dfe4c9033db/src/backend/distributed/utils/resource_lock.c#L1273-L1274

The change does not really affect lock since lock needs to be called in txn either way, but it affects TRUNCATE.

File: src/backend/distributed/commands/utility_hook.c [hunk 8]
```diff
@@ -468,6 +467,11 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 		PreprocessTruncateStatement((TruncateStmt *) parsetree);
 	}
 
+	if (IsA(parsetree, LockStmt))
+	{
+		PreprocessLockStatement((LockStmt *) parsetree, context);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/utility_hook.c:472]: minor: maybe add a comment (with proper formatting and wording): 
```SQL
/*
* PreprocessLockStatement may or may not lock the relations locally depending on whether the current node is in pg_dist_node. Still, there
* is no harm  to re-lock via standard process utility even if the node is in the metadata
```

File: src/backend/distributed/utils/resource_lock.c [hunk 9]
```diff
@@ -1196,3 +1216,164 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+			}
+		}
+		else if (!ShouldSyncTableMetadata(relationId))
+		{
+			continue;
+		}
+
+		distributedRelationList = list_append_unique_oid(distributedRelationList,
+														 relationId);
+		lockDistRelDescendantsList = lappend(lockDistRelDescendantsList,
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1310]: Instead of having two lists with the same size, should we create a new structure like
```
typedef struct LockRelation
{
Oid relationId;
bool inh;
} LockRelation;
```
Now, it is non-trivial to follow what we do with `lockDescendantsList`.

Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1310]: Instead of having two lists with the same size, should we create a new structure like
```
typedef struct LockRelation
{
Oid relationId;
bool inh;
} LockRelation;
```
Now, it is non-trivial to follow what we do with `lockDescendantsList`.

File: src/backend/distributed/utils/resource_lock.c [hunk 10]
```diff
@@ -1131,10 +1135,21 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
 			{
-				appendStringInfo(lockRelationsCommand, " IN %s MODE;\n", lockModeText);
+				appendStringInfo(lockRelationsCommand, lockTerminationString,
+								 lockModeText);
 				startedLockCommand = false;
 			}
 
+			/*
+			 * the user should not be able to trigger this codepath
+			 * with nowait = true or lockDescendants = flase since the
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1145]: typo: flase

Also, we generally start the multi-line comments with Capital letters

Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1145]: typo: flase

Also, we generally start the multi-line comments with Capital letters

File: src/backend/distributed/utils/resource_lock.c [hunk 11]
```diff
@@ -1104,22 +1114,16 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
 	 */
 	bool startedLockCommand = false;
 
+	/* Used to terminate partial lock commands */
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1117]: minor: we can move this declaration inside of the if statement where it is used.

Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1117]: minor: we can move this declaration inside of the if statement where it is used.

File: src/backend/distributed/utils/resource_lock.c [hunk 12]
```diff
@@ -1196,3 +1216,164 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+				{
+					referencingRelationIds = lappend_oid(referencingRelationIds,
+														 referencingRelationId);
+				}
+			}
+		}
+	}
+
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, referencingRelationIds)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1333]: why don't we immediately add referencingRelationIds to distributedRelationList earlier but defer here?

Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1333]: why don't we immediately add referencingRelationIds to distributedRelationList earlier but defer here?

File: src/backend/distributed/utils/resource_lock.c [hunk 13]
```diff
@@ -1196,3 +1216,164 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+		ErrorIfIllegallyChangingKnownShard(relationId);
+
+		/*
+		 * we want to prevent under privileged users to trigger establishing connections,
+		 * that's why we have this additional check. Otherwise, we'd get the errors as
+		 * soon as we execute the command on any node
+		 */
+		EnsureCanAcquireLock(relationId, lockMode);
+
+		if (get_rel_relkind(relationId) == RELKIND_VIEW)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1293]: minor simplification suggestion:

I think the rest of the loop can be moved to a separate function like:
```C

List *relationsToBeLocked = FindRelationsToBeLocked(relationId);
distributedRelationList = list_concat(distributedRelationList, relationsToBeLocked)
```

In general, we try to avoid more than `3` levels of indentation (if/while/switch etc) for keeping the code relatively easier to follow

Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1293]: minor simplification suggestion:

I think the rest of the loop can be moved to a separate function like:
```C

List *relationsToBeLocked = FindRelationsToBeLocked(relationId);
distributedRelationList = list_concat(distributedRelationList, relationsToBeLocked)
```

In general, we try to avoid more than `3` levels of indentation (if/while/switch etc) for keeping the code relatively easier to follow

File: src/test/regress/expected/distributed_locks.out [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+ERROR:  LOCK TABLE can only be used in transaction blocks
+-- Test acquiring lock inside procedure
+DO $$
+BEGIN
+LOCK dist_table IN ACCESS EXCLUSIVE MODE;
+END;
+$$;
+-- Test that when the user does not have the required permissions to lock
+-- the locks are not forwarded to the workers
+CREATE USER read_only_user;
```
Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:23]: Minor: These commands work on enterprise and hence avoids the NOTICE/HINT.

So, we typically do the following to prevent divergence of test outputs on community vs enterprise:
```SQL
SET client_min_messages TO ERROR;
 CREATE USER read_only_user;
RESET client_min_messages ;
```

Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:23]: Minor: These commands work on enterprise and hence avoids the NOTICE/HINT.

So, we typically do the following to prevent divergence of test outputs on community vs enterprise:
```SQL
SET client_min_messages TO ERROR;
 CREATE USER read_only_user;
RESET client_min_messages ;
```

File: src/test/regress/expected/distributed_locks.out [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+NOTICE:  not propagating CREATE ROLE/USER commands to worker nodes
+HINT:  Connect to worker nodes directly to manually create all necessary users and roles.
+GRANT SELECT ON distribute_lock_tests.dist_table TO read_only_user;
+SET ROLE read_only_user;
+SET citus.log_remote_commands TO ON;
+BEGIN;
+LOCK distribute_lock_tests.dist_table IN ACCESS EXCLUSIVE MODE;
+ERROR:  permission denied for schema distribute_lock_tests
+ROLLBACK;
+RESET citus.log_local_commands;
```
Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:33]: minor: we set log_remote_commands but reset log_local_commands

Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:33]: minor: we set log_remote_commands but reset log_local_commands

File: src/test/regress/expected/distributed_locks.out [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+BEGIN;
+LOCK local_table IN ACCESS EXCLUSIVE MODE;
+ROLLBACK;
+RESET citus.log_remote_commands;
+-- Cleanup local table
+DROP TABLE local_table;
+-- Test that setting the guc to 'on' will allow the lock from workers
+SET citus.allow_unsafe_locks_from_workers TO 'on';
+BEGIN;
+LOCK dist_table IN ACCESS EXCLUSIVE MODE;
```
Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:58]: Two quick notes:

1) we don't seem to have any successful tests from the coordinator?
2) it would be worthwhile to use all the lock modes, just to make sure we properly deparse all the lock modes 
```SQL

-- Try all valid lock options; also try omitting the optional TABLE keyword.
BEGIN TRANSACTION;
LOCK TABLE dist_table IN ACCESS SHARE MODE;
LOCK dist_table IN ROW SHARE MODE;
LOCK TABLE dist_table IN ROW EXCLUSIVE MODE;
LOCK TABLE dist_table IN SHARE UPDATE EXCLUSIVE MODE;
LOCK TABLE dist_table IN SHARE MODE;
LOCK dist_table IN SHARE ROW EXCLUSIVE MODE;
LOCK TABLE dist_table IN EXCLUSIVE MODE;
LOCK TABLE dist_table IN ACCESS EXCLUSIVE MODE;
ROLLBACK;
```

Comment (onderkalaci) [inline src/test/regress/expected/distributed_locks.out:58]: Two quick notes:

1) we don't seem to have any successful tests from the coordinator?
2) it would be worthwhile to use all the lock modes, just to make sure we properly deparse all the lock modes 
```SQL

-- Try all valid lock options; also try omitting the optional TABLE keyword.
BEGIN TRANSACTION;
LOCK TABLE dist_table IN ACCESS SHARE MODE;
LOCK dist_table IN ROW SHARE MODE;
LOCK TABLE dist_table IN ROW EXCLUSIVE MODE;
LOCK TABLE dist_table IN SHARE UPDATE EXCLUSIVE MODE;
LOCK TABLE dist_table IN SHARE MODE;
LOCK dist_table IN SHARE ROW EXCLUSIVE MODE;
LOCK TABLE dist_table IN EXCLUSIVE MODE;
LOCK TABLE dist_table IN ACCESS EXCLUSIVE MODE;
ROLLBACK;
```

File: src/test/regress/expected/isolation_acquire_distributed_locks.out [hunk 17]
```diff
@@ -0,0 +1,1031 @@
+Parsed test spec with 3 sessions
```
Comment (onderkalaci) [inline src/test/regress/expected/isolation_acquire_distributed_locks.out:1]: amazing isolation test coverage, thanks!

I think there is one more test I'd love to see is we directly/only lock citus_local_table  (not the view as well) and show that the workers are blocked.

Comment (onderkalaci) [inline src/test/regress/expected/isolation_acquire_distributed_locks.out:1]: amazing isolation test coverage, thanks!

I think there is one more test I'd love to see is we directly/only lock citus_local_table  (not the view as well) and show that the workers are blocked.

File: src/backend/distributed/utils/resource_lock.c (hunk 12 above)
Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1333]: https://github.com/citusdata/citus/blob/ac76db83e8a02e5c00923f24a18b2efa3e2acc91/src/backend/distributed/utils/resource_lock.c#L1243-L1265

File: src/backend/distributed/utils/resource_lock.c (hunk 11 above)
Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1117]: `lockTerminationString` is used as in a couple of places (inside the main loop and at the end of it) so outside the loop is the best place I could think of in terms of proximity to where it is used.

File: src/backend/distributed/utils/resource_lock.c (hunk 12 above)
Comment (onderkalaci) [reply src/backend/distributed/utils/resource_lock.c:1333]: So, 3 tables `tbl1`, `tbl2`, `tbl3`, where we have `tbl2 <- tbl1`

 When I do:
```SQL
LOCK tbl2, tbl3;
```

I'd expect the order to be `tbl2, tbl1, tbl3`. But, we seem to follow: `tbl2, tbl3, tbl1`.

Is that right? If so, why do you think we should insert `tbl3` in between?

Comment (gledis69) [reply src/backend/distributed/utils/resource_lock.c:1333]: What you suggest makes sense. I guess I misinterpreted :locking the referencing tables after" as:
* lock referencing tables after the all the tables that are explicitly specified.

instead of:
* lock referencing tables right after the table they reference.

I will switch the locking order to your suggestion.

File: src/backend/distributed/utils/resource_lock.c [hunk 18]
```diff
@@ -1196,3 +1245,169 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
...
+	if (!ClusterHasKnownMetadataWorkers() || !EnableMetadataSync || !EnableDDLPropagation)
+	{
+		return;
+	}
+
+	/* used to store the relations that will be locked */
+	List *distributedRelationRecordsList = NIL;
+
+	/* used to ensure uniqueness of the relations that will be locked */
+	List *uniqueDistributedRelationList = NIL;
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1308]: Although we have a significant improvement, I still find keeping two lists hard to follow, especially given one is unique the other is not.

Can't we implement uniqueness checks on distributedRelationRecordsList? See some examples via `TaskListMember` / `TaskListDifference` / `TasksEqual` etc on list operations specific for a type

**Review by onderkalaci — COMMENTED** [review]
I only have a single comment left, other than that, we are pretty close to merge.

File: src/backend/distributed/utils/resource_lock.c [hunk 19]
```diff
@@ -1196,3 +1245,169 @@ AcquireDistributedLockOnRelations(List *relationIdList, LOCKMODE lockMode)
 		SendCommandToWorker(nodeName, nodePort, lockCommand);
```
Comment (onderkalaci) [inline src/backend/distributed/utils/resource_lock.c:1245]: Ah, missed that. It is probably more accurate to use `SendMetadataCommandListToWorkerInCoordinatedTransaction()`. It internally uses `METADATA CONNECTION` which is useful to prevent edge cases in tx blocks. We should use metadata connection because we are acting on the distributed shell tables

```C

SendMetadataCommandListToWorkerInCoordinatedTransaction(list_make1(lockCommand))
```

One additional test could be the following, which sometimes might fail without the fix:
```SQL
BEGIN;
create table tb(a int);
select create_distributed_table('tb', 'a');
lock  tb in access share mode ;
```

**Review by onderkalaci — APPROVED** [review]
Looks ready to :shipit:

**Comment by hanefi** [comment]
How about the following changelog entry?

  * Adds support for `LOCK` commands

**Comment by hanefi** [comment]
I will use the following that suits our wording in changelog.

* Adds support for `LOCK` commands on distributed tables from worker nodes
